src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs

```diff
@@ -234,25 +235,27 @@ internal void ApplyHttpsDefaults(HttpsConnectionAdapterOptions httpsOptions)
     /// </summary>
     public void Load()
     {
-        if (_loaded)
+        if (_reloadToken is null || _reloadToken.HasChanged)
         {
-            // The loader has already been run.
-            return;
+            // Will update _reloadToken
+            _ = Reload();
         }
-        _loaded = true;
-
-        Reload();
 
         foreach (var action in EndpointsToAdd)
         {
             action();
         }
+
+        // If Load is called again, we don't want to rerun these
+        EndpointsToAdd.Clear();
     }
 
     // Adds endpoints from config to KestrelServerOptions.ConfigurationBackedListenOptions and configures some other options.
     // Any endpoints that were removed from the last time endpoints were loaded are returned.
     internal (List<ListenOptions>, List<ListenOptions>) Reload()
     {
+        _reloadToken = Configuration.GetReloadToken();
```

Comment:
```suggestion
        _reloadToken = ReloadOnChange ? Configuration.GetReloadToken() : NullChangeToken.Singleton;
```

We should not grab the token on load if `KestrelConfigurationLoader.ReloadOnChange` is false. That's what the is `DoesNotReloadOnConfigurationChangeByDefault` test is about. Normally it's true if you have a "default" host like `WebApplication`.

You'll need to add using `Microsoft.Extensions.FileProviders` too. Not the best namespace there.

Reply:
I didn't change the existing `IChangeToken` handling, so I believe that test is still passing.  I intentionally decoupled rebind-on-change from determine-if-reloading-would-be-redundant, but I'm open to discussion.